File: src/main/java/net/sjrx/intellij/plugins/systemdunitfiles/semanticdata/optionvalues/ServiceTypeOptionValue.java
package net.sjrx.intellij.plugins.systemdunitfiles.semanticdata.optionvalues;

import com.google.common.collect.ImmutableSet;

import java.util.Set;

/**
 * This validator is for the config_service_type validator.
 * <p></p>
 * This seems to use the values referenced in:
 * <p></p>
 * service_type_table ~ line 3990.
 */
public class ServiceTypeOptionValue implements OptionValueInformation {
  
  private static final Set<String> validOptions = ImmutableSet.of("simple", "forking", "oneshot", "dbus", "notify", "idle", "exec");
  
  @Override
  public Set<String> getAutoCompleteOptions() {
    return validOptions;
  }
  
  @Override
  public String getErrorMessage(String value) {
    if (validOptions.contains(value)) {
      return null;
    } else {
      return "Expected value " + value + " does not match one of the expected options: " + validOptions;
    }
  }
  
  @Override
  public String getValidatorName() {
    return "config_parse_service_type";
  }
}

File: src/main/kotlin/net/sjrx/intellij/plugins/systemdunitfiles/semanticdata/optionvalues/ServiceTypeOptionValue.kt
package net.sjrx.intellij.plugins.systemdunitfiles.semanticdata.optionvalues

import com.google.common.collect.ImmutableSet
import com.intellij.openapi.project.Project
import net.sjrx.intellij.plugins.systemdunitfiles.semanticdata.Validator

/**
 * This validator is for the config_service_type validator.
 *
 *
 * This seems to use the values referenced in:
 *
 *
 * service_type_table ~ line 3990.
 */
class ServiceTypeOptionValue : AbstractEnumOptionValue(validOptions, VALIDATOR_NAME) {

  companion object {
    private val validOptions: Set<String> = ImmutableSet.of("simple", "forking", "oneshot", "dbus", "notify", "idle", "exec")

    const val VALIDATOR_NAME = "config_parse_service_type"
    val validators = mapOf(Validator(VALIDATOR_NAME, "0") to ServiceTypeOptionValue())
  }
}
